theme-kit/router.ts

```diff
@@ -0,0 +1,147 @@
+/* eslint-disable @typescript-eslint/no-use-before-define,@typescript-eslint/naming-convention */
+
+import type { Culture } from '@prezly/sdk';
+import { notFound } from 'next/navigation';
+import type { ReactElement, ReactNode } from 'react';
+import UrlPattern from 'url-pattern';
+
+import { assertServerEnv } from '@/utils';
+
+import { api } from './api';
+
+export function route<
+    Pattern extends `/${string}`,
+    Match extends ExtractPathParams<Pattern>,
+    Props = Match,
+>(pattern: Pattern, loader: Loader<Match, Props>) {
+    const urlPattern = buildUrlPattern(pattern);
+
+    return {
+        async match(path: string, getDefaultLocale, isEnabledLocale) {
+            const matched = urlPattern.match(path);
+
+            if (!matched) {
+                return undefined;
+            }
+
+            if (matched.locale && !(await isEnabledLocale(matched.locale))) {
+                return undefined;
+            }
+
+            if (!matched.locale) {
+                matched.locale = await getDefaultLocale();
+            }
+
+            const { match: pageMatch, resolveLocale, Layout, default: Page } = await loader();
+
+            const pageMatched = pageMatch ? await pageMatch(matched) : matched;
+
+            if (!pageMatched) {
+                return undefined;
+            }
+
+            if (resolveLocale) {
+                return {
+                    match: {
+                        ...pageMatched,
+                        locale: await resolveLocale(pageMatched),
+                    },
+                    Layout,
+                    Page,
+                };
+            }
+
+            return {
+                match: { ...pageMatched, locale: matched.locale ?? (await getDefaultLocale()) },
+                Layout,
+                Page,
+            };
+        },
+    } as Route<Match, Props | Match>;
+}
+
+type Route<Match, Props> = {
+    match(
+        path: string,
+        getDefaultLocale: () => Promise<string>,
+        isEnabledLocale: (locale: string) => Promise<boolean>,
+    ): Promise<
+        | {
+              match: Match;
+              Layout?: ServerComponentType<Props>;
+              Page: ServerComponentType<Props>;
+          }
+        | undefined
+    >;
+};
+
+export async function match(path: string | string[], routes: Route<any, any>[]) {
+    assertServerEnv('match');
+
+    const { contentDelivery } = api();
+
+    async function getDefaultLocale() {
+        return (await contentDelivery.language())!.locale.code;
+    }
+
+    async function isEnabledLocale(locale: string) {
+        return contentDelivery.language(locale).then(Boolean);
+    }
+
+    const stringPath = typeof path === 'string' ? path : `/${path.join('/')}`;
+    const matches = await Promise.all(
+        routes.map((r) => r.match(stringPath, getDefaultLocale, isEnabledLocale)),
+    );
```

Comment:
Route matching is done with parallel async execution to evaluate the matched route URLs against the API data without waiting for each response sequentially.

Reply:
I wonder if we could simplify this whole thing by using `middleware`.
The only real conflict we have is `/[locale]` vs `/[slug]`.

What if we define a `/story/[slug]` route, then have a Next middleware function process all requests that match `/[locale|slug]` and based on the checks would rewrite to `/story/[slug]`?

Reply:
> and based on the checks would rewrite to `/story/[slug]`?

I don't like this particular aspect of your suggestion. I'd not want to do internal rewriting magic to keep things directly mapped between the public URL and internal structure.

But it's a great point that having parts of this unloaded to a middleware may help. I'll explore it. 

Thanks!